### DISCUSSION TITLE
How to run mitmproxy in Heroku?

### DISCUSSION THREAD

**Discussion by sandeshth** [mitmproxy/mitmproxy#6758] [Q&A]
I am using mitmproxy with Python. Python version is 3.11.0 and mitmproxy version is 10.2.0. When I install mitmproxy, it installed other dependencies such as 
```
mitmproxy-macos==0.5.1
mitmproxy_rs==0.5.1
urwid-mitmproxy==2.1.2.1
```
When I run in my local environment, it works fine but when I tried to run in Heroku, it returns an error. The error looks like this 

```
Traceback (most recent call last):
  File "/app/.heroku/python/bin/mitmdump", line 5, in <module>
     from mitmproxy.tools.main import mitmdump 
ModuleNotFoundError: No module name 'mitmproxy.tools'
```

When I check build log in Heroku, the package is installed properly with same version that I have in my local environment. 

After running Python main File, I am running mitmproxy as follows:
```
subprocess.Popen(['mitmdump', '-s', 'filepath']) 
```

**Comment by rosydawn6** [comment]
Can you verify that you are using the correct version of python and not for example the default/stock system python unless they are one in the same?

Once confirmed you should start the python interpreter from the VM and try to invoke the command that is causing the issue as it may be necessary to set the python path :

```
import sys
sys.path
from mitmproxy.tools.main import mitmdump

```

**Reply by sandeshth** [reply]
I am trying to use mitmproxy with following example 
https://scrapfly.io/blog/how-to-block-resources-in-selenium/

I have main file that will start the python and want to run mitmdump -s block.py using subprocess.popen. It runs fine in my local but not in Heroku.

**Reply by sandeshth** [reply]
I logged into Heroku and started the python interpreter. 
When I typed `import mitmproxy`, it did not return any error. 
But when I typed `from mitmproxy.tools.main import mitmdump`, it returned following error 

```
Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'mitmproxy.tools'    
```

**Reply by rosydawn6** [reply]
It looks like the installation and directory structure is different on the Heroku server or there is some type of visibility/version mismatch.  Which python version are you starting on the server?  Is it the same python that has mitmproxy available?  What does the following give?
```
import mitmproxy
dir(mitmproxy)
mitmproxy.__version__
```

**Reply by sandeshth** [reply]
On my local environment, the python version is `3.11.0` and on Heroku environment, it is `3.11.2`. This url https://docs.mitmproxy.org/stable/overview-installation/#:~:text=(we%20require%20at%20least%203.10). mentioned about atleast  `3.10`. So, I think it satisfied the condition, I think 

I started the bash cell in Heroku, then typed 

```
python

It started the python terminal 

import mitmproxy 
dir(mitmproxy)
['__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__']

mitmproxy.__version__

returned an error 

Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
AttributeError: module 'mitmproxy' has no attribute '__version__'
```

Just adding these two links wondering that might be helpful to you 
```
https://github.com/mitmproxy/mitmproxy/discussions/5121

https://github.com/mitmproxy/mitmproxy/discussions/5255#:~:text=from%20mitmproxy.tools.main%20import%20mitmdump%0A%0Amitmdump(args%3D%5B%22%2Ds%22%2C%20%22myaddon.py%22%5D)
```

I tried second link and implemented in my local environment but it started proxy but did not proceed to next line.

**Reply by sandeshth** [reply]
I have same python version in Heroku as well `3.11.0` and still it returned same error.

**Comment by rosydawn6** [comment]
Can you confirm that mitmproxy module is visible from within the Heroku environment?  Did you install mitmproxy via this method https://docs.mitmproxy.org/dev/overview-installation/#installation-from-the-python-package-index-pypi
What do the below return?
```
from mitmproxy import exceptions
from mitmproxy import version
```
